packages/frontend/src/hooks/useFunnelChartConfig.ts

```diff
@@ -0,0 +1,178 @@
+import {
+    FunnelChartDataInput,
+    isField,
+    isMetric,
+    isTableCalculation,
+    type ApiQueryResults,
+    type CustomDimension,
+    type Dimension,
+    type FunnelChart,
+    type ItemsMap,
+    type Metric,
+    type TableCalculation,
+    type TableCalculationMetadata,
+} from '@lightdash/common';
+import { useEffect, useMemo, useRef, useState } from 'react';
+import { type FunnelSeriesDataPoint } from './echarts/useEchartsFunnelConfig';
+
+type FunnelChartConfig = {
+    validConfig: FunnelChart;
+
+    fieldId: string | null;
+    maxValue: number;
+    selectedField: Metric | TableCalculation | undefined;
+    fieldChange: (fieldId: string | null) => void;
+
+    dataInput: FunnelChartDataInput;
+    setDataInput: (dataInput: FunnelChartDataInput) => void;
+
+    data: FunnelSeriesDataPoint[];
+};
+
+export type FunnelChartConfigFn = (
+    resultsData: ApiQueryResults | undefined,
+    funnelChartConfig: FunnelChart | undefined,
+    itemsMap: ItemsMap | undefined,
+    dimensions: Record<string, CustomDimension | Dimension>,
+    numericFields: Record<string, Metric | TableCalculation>,
+    colorPalette: string[],
+    tableCalculationsMetadata?: TableCalculationMetadata[],
+) => FunnelChartConfig;
+
+const useFunnelChartConfig: FunnelChartConfigFn = (
+    resultsData,
+    funnelChartConfig,
+    itemsMap,
+    dimensions,
+    numericFields,
+    colorPalette,
+    tableCalculationsMetadata,
+) => {
+    const [fieldId, setFieldId] = useState(funnelChartConfig?.fieldId ?? null);
+
+    const [dataInput, setDataInput] = useState(
+        funnelChartConfig?.dataInput ?? FunnelChartDataInput.ROW,
+    );
+
+    // The value at the top of the funnel.
+    const maxValue = useRef(0);
```

Comment:
why do we have this?

Reply:
This is the value that the funnel 'starts' with. Percentages in later stages are based on this number. I was surprised how bad a job echarts does here -- we have to calculate percentages ourselves to get it right. 

I will make the comment more clear